From 45eaee67fe5d78fe3798da6192297ea1c50079f0 Mon Sep 17 00:00:00 2001
From: David Vrabel <david.vrabel@citrix.com>
Date: Thu, 8 May 2014 18:22:58 +0100
Subject: [PATCH 2/3] x86/nmi: enable local irqs in wait_for_nmis()

wait_for_nmis() runs for 10 ticks with local interrupts disabled,
which is quite a long time.

We occasionally see serial output (particularly with virtual serial
ports such as HP's iLO) stopping on the "Testing for NMIs" line.  We
think the long delay in handling the serial interrupt is the cause.

Signed-off-by: David Vrabel <david.vrabel@citrix.com>
---
 xen/arch/x86/nmi.c |    7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/xen/arch/x86/nmi.c b/xen/arch/x86/nmi.c
index 18d3820..07488da 100644
--- a/xen/arch/x86/nmi.c
+++ b/xen/arch/x86/nmi.c
@@ -111,7 +111,14 @@ int nmi_active;
 
 static void __init wait_for_nmis(void *p)
 {
+    unsigned long flags;
+
+    local_save_flags(flags);
+    local_irq_enable();
+
     mdelay((10*1000)/nmi_hz); /* wait 10 ticks */
+
+    local_irq_restore(flags);
 }
 
 int __init check_nmi_watchdog (void)
-- 
1.7.10.4

